pkg/dao/db.go

```diff
@@ -1,20 +1,56 @@
 package dao
 
-import "database/sql"
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+
+	_ "github.com/lib/pq"
+)
 
 var db *sql.DB
 
-func Connect() (*sql.DB, error) {
+func Connect(dbHost, dbUser, dbPassword, dbName, sslMode string) (*sql.DB, error) {
 	if db != nil {
 		return db, nil
 	}
 	//connection logic
-	return nil, nil
+	connStr := fmt.Sprintf("host=%v user=%v password=%v dbname=%v sslmode=%v", dbHost, dbUser, dbPassword, dbName, sslMode)
+
+	// sql.Open doesn't initialize the connection immediately
+	dbInstance, err := sql.Open("postgres", connStr)
+
+	// an error can happen here if the connection string is invalid
+	if err != nil {
+		return nil, err
+	}
+
+	// an error happens here if we cannot connect
+	if err = dbInstance.Ping(); err != nil {
+		return nil, err
+	}
+
+	// assign db variable declared above
+	db = dbInstance
+	return dbInstance, nil
 }
 
 func Disconnect() error {
-	if nil != db {
+	if db != nil {
 		return db.Close()
 	}
 	return nil
 }
+
+func DoInitialSetup() error {
+	if db == nil {
+		return errors.New("cannot setup database, must call Connect() first")
+	}
+	if _, err := db.Exec("CREATE TABLE IF NOT EXISTS sdkVersionForClient(clientId varchar(10) not null primary key, version varchar(40) not null, event_time timestamp with time zone)"); err != nil {
+		return err
+	}
```

Comment:
@darahayes @wtrocki  I actually found out a much better way for the knowledge base.

Now that we're using Postgres, much more complex aggregations are possible. ElasticSearch had limitations...

We can create a view that contains the latest entries for the "mobileAppMetrics" table (created below) and use that in Grafana.

No need for 2nd insertion/update operation nor a separate table.

But we need to create a view:

```sql
create view c as select entry.clientId, entry.event_time, entry.data from mobileappmetrics entry inner join (select clientId, max(event_time) as latestEntryTime from mobileappmetrics group by clientId) latestEntry on entry.clientId = latestEntry.clientId and entry.event_time = latestEntryTime;
```

Doing `select * from clientKnowledgeBase` would return the data de-duplicated per client id.

This, however, means that we should use this table only for our metrics, not custom metrics as we would want this view to have entries only for the metrics that needs de-duplication.

I gave this one a try on the POC but haven't checked in anything.

Reply:
@aliok Sounds really nice. Perhaps you should push it to a branch so we can look at it!